server/util.go

```diff
@@ -62,18 +66,71 @@ func PersistentPreRunEFn(context *Context) func(*cobra.Command, []string) error
 
 // add server commands
 func AddCommands(
-	rootCmd *cobra.Command,
-	appState GenAppState, appCreator AppCreator,
-	context *Context) {
+	ctx *Context, cdc *wire.Codec,
+	rootCmd *cobra.Command, appInit AppInit,
+	appCreator AppCreator) {
 
-	rootCmd.PersistentFlags().String("log_level", context.Config.LogLevel, "Log level")
+	rootCmd.PersistentFlags().String("log_level", ctx.Config.LogLevel, "Log level")
 
 	rootCmd.AddCommand(
-		InitCmd(appState, context),
-		StartCmd(appCreator, context),
-		UnsafeResetAllCmd(context),
-		ShowNodeIDCmd(context),
-		ShowValidatorCmd(context),
+		InitCmd(ctx, cdc, appInit),
+		StartCmd(ctx, appCreator),
+		UnsafeResetAllCmd(ctx),
+		ShowNodeIDCmd(ctx),
+		ShowValidatorCmd(ctx),
 		version.VersionCmd,
 	)
 }
+
+//___________________________________________________________________________________
+
+// append a new json field to existing json message
+func AppendJSON(cdc *wire.Codec, baseJSON []byte, key string, value json.RawMessage) (appended []byte, err error) {
+	var jsonMap map[string]json.RawMessage
+	err = cdc.UnmarshalJSON(baseJSON, &jsonMap)
+	if err != nil {
+		return nil, err
+	}
+	jsonMap[key] = value
+	bz, err := wire.MarshalJSONIndent(cdc, jsonMap)
+	return json.RawMessage(bz), err
+}
+
+// https://stackoverflow.com/questions/23558425/how-do-i-get-the-local-ip-address-in-go
+// TODO there must be a better way to get external IP
+func externalIP() (string, error) {
```

Comment:
This won't work too well for NAT'ed / multi-IP / Docker / any complex network setup, which seems likely for validators.

An external API - e.g. [this](https://ipinfo.io/) - would be more reliable, but I'm not sure that's a dependency we want to introduce.

I think we should at least allow manual IP specification, and maybe require it always.

Reply:
good idea for manual IP flag 👍

Reply:
But I'd prefer not to depend on an external service for IP retrieval though - so not into using that website

Reply:
Agreed, and it still wouldn't be infallible, a manual flag is better (and maybe if we determine the IP automatically we should print it so the user knows which one we picked).